### DISCUSSION TITLE
Related parts - purpose unclear

### DISCUSSION THREAD

**Discussion by turist79** [inventree/InvenTree#8416] [General]
Before coming to the Inventree i have tried different apps (Dolibarr, Odoo and Triton being all-in-one ERPs and Openboxes as WMS-only). I hope it will be Ok if i will from time to time drop some ideas based on features i have found useful in those apps.

One of things i miss in Inventree are Catalogs from Openboxes  https://help.openboxes.com/article/319-create-a-product-catalog
They provide the way to group different items together with many to many relation (catalog can include many parts, one part can be referenced by many catalogs). Parts may be related to some specific project, equipment type etc.

In Inventree this may be achieved probably by creating separate category with virtual parts representing the catalogs and assigning all the required parts to it via "related parts". But i`m not sure that this wont backfire later when some functionality will be introduced for the related parts - so i would like to know if there are any plans regarding further development of related parts function?

The other thing from Openboxes is Product association https://help.openboxes.com/article/320-create-product-associations
Currently the use of related parts is somewhat limited in my opinion since there is even no simple description field for the relation which could provide the information on the naure of the relation. This looks like good place to keep more structured information about part relations.
Since substitutes are handled on BOM level, variants on the part level - so being accessory, related consumable (like thermal compaund for cpu//mainboard for example)?

**Comment by SchrodingersGat** [comment]
>  I hope it will be Ok if i will from time to time drop some ideas based on features i have found useful in those apps.

Yes, 100% - any input / suggestions / etc are always welcome :)

> One of things i miss in Inventree are Catalogs from Openboxes 

We support "tags" for parts, however currently these are not integrated into the interface. Tags can be used by plugins, but the concept has not been fully fleshed out and made available in the UI. I think that this would maybe achieve what you are looking for here? You can assign multiple "tags" to a particular part, and any other part which has a matching tag could be easily matched.

> Currently the use of related parts is somewhat limited in my opinion since there is even no simple description field for the relation which could provide the information on the naure of the relation. This looks like good place to keep more structured information about part relations.

Adding a "description" field to the `PartRelation` table would be a good idea, I think. That's what you are going for here?

**Comment by SergeoLacruz** [comment]
I am also thinking about a "relation type" or description like "mates with" for connectors or "is replaced by" for old parts. I tried to implement it but could not finish it so far because I still need to learn how the details work.

**Reply by turist79** [reply]
> I am also thinking about a "relation type" or description like "mates with" for connectors or "is replaced by" for old parts. I tried to implement it but could not finish it so far because I still need to learn how the details work.

May i ask how far have you moved in this attempt? I`m currently looking into example plugins and some plugins made by community to get the understanding of what plugin engine is capable of and what is impossible to implement with plugins.

=

As far as i understood by now it is possible to integrate additional sections (tabs) with extra information to the existing object like (part, stock item, sales order, purchase order etc), but it looks like it is **not** possible to change how the already existing elements like fields, lists, forms (for example, part creation, stock transfer dialog etc) look like and add additional fields there. So in this case it is not possible to extend the realtion record with some aditional fields which will be visible in the same 'related parts' list as additional columns, assignable and editable within the same interface as the relation itself.

In terms of storing some data, plugins are limited to storing small amount of it  as json inside Plugin Metadata of the specific object.
In terms of manipulation and display of this data it will always be some separate UI section.

Everything above is my first impression which could be totally wrong, i have very limited python experience (mostly web scrapping) and even less django :)

**Reply by matmair** [reply]
There are a lot of wrong sugestionas here - please read the plugin docs - [this](https://docs.inventree.org/en/stable/extend/how_to_plugin/) is a good starting page.

**Comment by SergeoLacruz** [comment]
A plugin is not needed here. There is already a PartRelated class with two fields, part_1 and part_2. All it needs is an additional CharField. This also needs to go into the serializers, api, admin. For someone skilled in Django this is probably a simple task. Then we need to make it editable in CUI and also in PUI. I never touched PUI so far...

**Reply by SchrodingersGat** [reply]
@SergeoLacruz check out https://github.com/inventree/InvenTree/pull/8524 - as this is a very simple addition it would be a good PR for you to study if you do have a chance to contribute in the future :)

**Reply by SergeoLacruz** [reply]
This looks understandable. Thanks for the hint. I think there are changes missing for the CUI.

**Reply by matmair** [reply]
https://github.com/inventree/InvenTree/pull/8384

**Reply by SchrodingersGat** [reply]
New features are only being supported in the new UI
